### DISCUSSION TITLE
Dockerfile: Cannot find module 'agentkeepalive' on version 9.20

### DISCUSSION THREAD

**Discussion by benjibenoit** [directus/directus#16327] [Archive]
### Describe the Bug

I created my own dockerfile like this:
![image](https://user-images.githubusercontent.com/32774804/200044066-b96c136f-1642-4783-9ad9-fd7cb8124d4d.png)
When I build it, I get the error below:
![image](https://user-images.githubusercontent.com/32774804/200044119-c95e0d6b-9da0-4387-980a-6619793b4afe.png)

I could also add that it works very well with the 9.19 and all the ones below

### To Reproduce

Create dockerfile with 
![image](https://user-images.githubusercontent.com/32774804/200044163-f7c12efd-1155-4174-bdc4-e19fa6da9b34.png)
And build it on Mac

### What version of Directus are you using?

9.20

### What version of Node.js are you using?

16

### What database are you using?

postgres

### What browser are you using?

Chrome

### How are you deploying Directus?

Docker

**Comment by azrikahar** [comment]
`agentkeepalive` is a nested dependency of `sqlite3`, so something doesn't seem right here 🤔

```
`-- sqlite3@5.1.2
  `-- node-gyp@8.4.1
    `-- make-fetch-happen@9.1.0
      `-- agentkeepalive@4.2.1
```

I'm not able to reproduce this on Ubuntu:

https://user-images.githubusercontent.com/42867097/200053835-c1354a67-9452-435f-8739-b783c44bd58c.mp4

But it feels like your local `directus/directus:9.20` image is (somehow) incomplete/corrupt. 

Would you mind testing to remove and re-pulling that image again? or maybe running the base `directus/directus:9.20` image, not a custom Dockerfile, and run `docker exec <your-container-name> npm list agentkeepalive` to verify whether that package exists similar to what I did in the screen capture above.

it would also be helpful if you can try it on another machine as well, just to rule out any machine-specific issues.

**Reply by benjibenoit** [reply]
I just tried everything. Delete all my images. Run version 9, or 9.20. Nothing works.
I can't list the npm packages either because it doesn't run:
![image](https://user-images.githubusercontent.com/32774804/200113663-2ac6ab93-f4df-4c4a-b192-37029ed10449.png)
I don't know if this can help, but I remind you that the problem only occurs on MacOs (it works very well for my colleagues with Windows). However as it is docker, I do not see why it would be a problem.
As a reminder: version 9.19 works without any problem.

**Reply by azrikahar** [reply]
> However as it is docker, I do not see why it would be a problem.

I was under the same impression as well! Hence I implicitly put that as an afterthought, but I should have still mention that nonetheless 👍

> the problem only occurs on MacOs (it works very well for my colleagues with Windows). 

Thanks for sharing the additional detail regarding your colleagues on Windows. I think the only potential difference now would be your colleague and I ended up using the x86 architecture docker image, whereas you might be using the arm architecture docker image 🤔 just to confirm, is your Mac on an arm chip (M1/M2)?

**Reply by benjibenoit** [reply]
Absolutely, I currently have an Apple M1.

**Reply by azrikahar** [reply]
Thanks for confirming! However, unfortunately I've also tried it on a Raspberry Pi 4b which also runs on ARM architecture without any issue, so I'm also rather confused what else could be going on here.

Here's some commands I tested to verify it is indeed using the arm64 directus image:

![](https://user-images.githubusercontent.com/42867097/200741243-98bcd435-5403-4e32-88bd-d7abfba3e80a.png)

The docker image inspect command used for easier copy/paste: `docker image inspect directus/directus:9.20.4 | grep Architecture`

Perhaps you can try running the `docker image inspect` command as an additional confirmation if it's pulling the arm64 image? Apart from that, I'm not sure what could be the culprit here as I do assume/believe there shouldn't be any disparity when it comes to docker images (except architecture), unless we can have another additional sanity check from someone else with Apple M1.

**Reply by benjibenoit** [reply]
Hey,
I just tested your commands. I have copied your entire docker compose, which is still impossible to start.
Here are the two commands I was able to execute as requested:
![image](https://user-images.githubusercontent.com/32774804/200802579-f4c82d8e-d147-4373-9754-36f4110c5c8c.png)
![image](https://user-images.githubusercontent.com/32774804/200802605-1d5f2447-f077-4596-9038-ab8bb410a852.png)
The question I have is what is the difference between 9.19 and 9.20 that makes this one not work anymore. Because as soon as I switch back to 9.19 everything works perfectly.

**Reply by azrikahar** [reply]
Just realised the docker compose was over simplified and missing `KEY` and `SECRET`, so do make sure those are present in your compose file as well.

```yaml
services:
    directus:
        image: directus/directus:9.20
        container_name: directus
        restart: unless-stopped
        ports:
            - 8055:8055
        environment:
            - KEY=key
            - SECRET=secret
            - ADMIN_EMAIL=admin@example.com
            - ADMIN_PASSWORD=password
            - LOG_STYLE=raw
```

I would also recommend setting `LOG_STYLE=raw` in case there's any additional error logs that went missing due to a known bug #16284.

> The question I have is what is the difference between 9.19 and 9.20 that makes this one not work anymore. 

AFAIK there was issue with the build pipeline for `9.20.0`, which was then resolved. 

There was a somewhat substantial dependencies update back in `9.19.0`, but not `9.20.0`, so unfortunately I'm also out of idea on why this particular package can be missing in your docker image (which in and of itself is very surprising to me with my current understanding of Docker).

**Reply by benjibenoit** [reply]
Nothing works. The KEY and SECRET don't change anything because the container doesn't even start and crashes directly. I have retested and I confirm that the bug is from 9.20.x
Same for the logs, it doesn't display any data because the container doesn't start.
Have you changed a dependency that could cause this? The error is "Module not found", is the docker build done correctly with all dependencies?
https://github.com/nvm-sh/nvm/issues/2438 Maybe this issue can help you ?

**Reply by azrikahar** [reply]
> Have you changed a dependency that could cause this? The error is "Module not found", is the docker build done correctly with all dependencies?

> https://github.com/nvm-sh/nvm/issues/2438 Maybe this issue can help you ?

Unfortunately I'm not aware of any dependency changes that could cause this, particularly since there has yet to be another similar report to serve as another reference point to potentially spot what could be the actual issue here.

From what I can tell at the moment, the build should be correct or else there would be more reports on it if it's not, but I totally understand that is not the case for you so this is a weird conundrum. The linked issue does seem to describe some of the potential causes, but AFAIK the build chain never uses `nvm` nor is it using `npm@7.x`, especially since #13455 specifically ensured the switch to npm `8.x` for Directus `9.12.0` and beyond.

@paescuj pardon for the sudden tag here, but do you by any chance have seen such issue before? 🤔 I'm not sure could the changes to the release workflow circa `9.20.0` (such as #16141) can have any impact which seemingly only happens to Apple M1 even with the official build Docker image?

**Reply by paescuj** [reply]
I cannot imagine that this problem is in any way related to #16141 and also don't see any other changes that could be responsible for it at short notice.

Just to be sure: So 9.19.2 is working on your side? Have you tested it with all the versions (9.20.0 - 9.20.4) since this release or just with the latest (9.20.4)?

Btw. I think it's super weird that you can't even run `npm list` in the container (https://github.com/directus/directus/discussions/16327#discussioncomment-4062890)... What happens if you run something like this: `docker run --rm directus/directus:9.20.4 ls node_modules/agentkeepalive`?

**Reply by benjibenoit** [reply]
Hi,
> Just to be sure: So 9.19.2 is working on your side? Have you tested it with all the versions (9.20.0 - 9.20.4) since this release or just with the latest (9.20.4)?

I just retested 9.19.2, without any problems. I then tested from 9.20.1 to 9.20.4, none of them works

> Btw. I think it's super weird that you can't even run `npm list` in the container ([#16327 (reply in thread)](https://github.com/directus/directus/discussions/16327#discussioncomment-4062890))... What happens if you run something like this: `docker run --rm directus/directus:9.20.4 ls node_modules/agentkeepalive`?

No npm command seems to work but the ls command does, here is the result:
![image](https://user-images.githubusercontent.com/32774804/201314495-00989ad7-a7ae-4f42-a9be-0ae1977c3ced.png)

Thanks again both of you for your time in searching for this error!

**Reply by paescuj** [reply]
Thanks @shankiflang! Hmm, really strange... Just built the ARM image locally and couldn't spot any abnormalities during the build process. I wonder if this issue is somehow related to M1/M2, would be interesting to see if it does not work for other users of this platform as well.

May I ask you to also run the following commands on your computer? Just to be sure about the versions:
```bash
docker run --rm directus/directus:9.19.2 node -v
docker run --rm directus/directus:9.19.2 npm -v
docker run --rm directus/directus:9.19.2 npm list agentkeepalive

docker run --rm directus/directus:9.20.4 node -v
# I expect 'npm' to work when not in /directus
docker run --rm -w / directus/directus:9.20.4 npm -v
```

**Reply by benjibenoit** [reply]
![image](https://user-images.githubusercontent.com/32774804/201326997-4359e1ca-5c21-4186-8e41-93cebbaa3dd5.png)

**Reply by paescuj** [reply]
Thanks! Sorry for the many requests, but can you also run this?:
```bash
docker run --rm directus/directus:9.19.2 uname -a
docker run --rm directus/directus:9.19.2 apk --print-arch
docker run --rm directus/directus:9.20.4 uname -a
docker run --rm directus/directus:9.20.4 apk --print-arch
docker run --rm directus/directus:9.20.4 ls /usr/local/lib/node_modules/npm/node_modules/agentkeepalive

docker image ls -a | awk '$1 == "directus/directus" { print $3 }' | xargs docker image inspect | grep -A2 'RepoTags\|Architecture'
```

**Reply by benjibenoit** [reply]
here are the 1st commands :
![image](https://user-images.githubusercontent.com/32774804/201362905-060ad7b4-4934-407d-93a9-39e8d2085979.png)

And the last one:
![image](https://user-images.githubusercontent.com/32774804/201362914-cda44d17-7f4d-48de-9bca-26186569eddb.png)
![image](https://user-images.githubusercontent.com/32774804/201362924-78987cb2-8ebf-46c4-a08d-9cc355f8a08f.png)

**Reply by benjibenoit** [reply]
(I deleted my old comment because it was in the wrong place, I just recreated it here to keep a better timeline.)
Here is the response to the command :
![image](https://user-images.githubusercontent.com/32774804/201363397-5bda625d-fb45-4dfd-8d65-56df41107a2e.png)

**Reply by paescuj** [reply]
Okay, those things are looking fine. I've added this command to the comment a few minutes ago, can you also run it?:
```bash
docker run --rm directus/directus:9.20.4 ls /usr/local/lib/node_modules/npm/node_modules/agentkeepalive
```

**Edit:**
Thanks for the addition!

There we have the issue: The directory `agentkeepalive` should actually be there. The question remains as to why this is missing. I've just tested it on another M1 machine and the problem did not occur there, so it almost seems like it's an issue specifically with your machine (maybe some corrupt layers...).

Here's the list of the layers of `directus/directus:latest` from the other machine:
```
                "sha256:5d3e392a13a0fdfbf8806cb4a5e4b0a92b5021103a146249d8a2c999f06a9772",
                "sha256:dae78dd4bc813fada04b19128ca13a79c6ccbbb27631fa3c6b2acba63633790e",
                "sha256:aa497f92dd156e7c7dc84ccc8ed74311f103fc7779e524fda8b4fb1ebf197383",
                "sha256:c1b87b5776c6154251f0988f206c49c660792dbcab7efd8c379510d685cefeea",
                "sha256:47fc9e9138807d9bfa492d19ba8560c341d92965ae18d696e0037aa7cd809a75",
                "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef",
                "sha256:1e102c6088930262376ed25bf7fd7659694f336a718ec86d6e381006a0c886b8",
                "sha256:6e1faa58d5491ede52af9ece9a5e7d42690eb8f4970cf7fff9f44f07377d88a2",
                "sha256:5c8e83465e1276ffcd042aa68f452666d68ca588a3ebe53df59131a774d217a3"
```

You might want to compare the layers with your image using something like `docker inspect directus/directus:latest | awk '/Layers/,/]/'` just to be sure.

Additionally, I'd suggest you to do a Docker system clean-up on your machine to get rid of any corrupt layers that may be present.

If you have no important containers / images, etc. you can simply run:
```bash
# ⚠️ This removes all containers, networks, images and build cache
docker ps -q | xargs docker stop
docker system prune -a
```
Otherwise, please make sure to remove at least all Directus images and other images where the base image is `node:16-alpine`.

Afterwards, try to spin up a Directus container again (🤞).

**Reply by benjibenoit** [reply]
> If you have no important containers / images, etc. you can simply run:
> 
> ```shell
> # ⚠️ This removes all containers, networks, images and build cache
> docker ps -q | xargs docker stop
> docker system prune -a
> ```

Just work perfectly after that.
![image](https://user-images.githubusercontent.com/32774804/201437032-0fbdabe9-f67d-4759-8556-34df49c7a2bd.png)

Many thanks again for a problem that was simply related to me.

**Reply by paescuj** [reply]
You're very welcome! Glad we were able to solve the problem and it's not a general thing! 👍

**Reply by azrikahar** [reply]
@paescuj thank you very much for helping out! Really appreciate it 🙌

**Reply by paescuj** [reply]
> @paescuj thank you very much for helping out! Really appreciate it 🙌

You're welcome 😃🙌
